buffer/src/test/java/io/netty/buffer/ByteBufUtilTest.java

```diff
@@ -542,6 +539,25 @@ public void testIsTextWithInvalidIndexAndLength() {
         }
     }
 
+    @Test
+    public void testUtf8BytesIsExact() {
+        final String s = "Some UTF-8 like äÄ∏ŒŒ";
+        checkUtf8Bytes(s);
+    }
+
+    @Test
+    public void testUtf8BytesWithSurrogatesIsExact() {
+        final String s = "a\uD800\uDC00b";
+        checkUtf8Bytes(s);
+    }
+
```

Comment:
also add the following test cases:
- includes a non-surrogate pair which results in 3 bytes (e.g. `'\uE000'`)
- includes a non-surrogate pair non-ascii (e.g. `(char) 0x81`)

Reply:
please also add update `testWriteUtf8Surrogates` to include the new method:

```java
assertEquals(buf.readableBytes(), ByteBufUtil.utf8Bytes(surrogateString));
```